cmd/registry/breaking-change-detector/breaking_change_detector.go

```diff
@@ -0,0 +1,212 @@
+
+	"github.com/apigee/registry/rpc"
+)
+
+type detectionPattern struct {
+	detectionWeight    int
+	PositiveMatchRegex *regexp.Regexp
+	NegativeMatchRegex *regexp.Regexp
+}
+
+var (
+	unsafeAdds = []detectionPattern{
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(components.)+(.|)+(schemas)+(.)+(required)"),
+		},
+	}
+
+	unsafeDeletes = []detectionPattern{
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(components.)+(.|)+(schemas)+(.)"),
+			NegativeMatchRegex: regexp.MustCompile("(components.)+(.|)+(schemas)+(.)+(required)"),
+		},
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(paths.)+(.|)"),
+			NegativeMatchRegex: regexp.MustCompile("((paths.)+(.|)+(tags))+(|)+((paths.)+(.|)+(description))"),
+		},
+	}
+
+	unsafeMods = []detectionPattern{
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(components.)+(.|)+(schemas)+(.|)+(type)"),
+		},
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(paths.)+(.|)+(type)"),
+			NegativeMatchRegex: regexp.MustCompile("((paths.)+(.|)+(tags))+(|)+((paths.)+(.|)+(description))"),
+		},
+	}
+
+	safeAdds = []detectionPattern{
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(info.)+(.)"),
+		},
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(tags.)+(.)"),
+		},
+	}
+
+	safeDeletes = []detectionPattern{
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(info.)+(.)"),
+		},
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(tags.)+(.)"),
+		},
+	}
+
+	safeMods = []detectionPattern{
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(info.)+(.)"),
+		},
+		{
+			detectionWeight:    1,
+			PositiveMatchRegex: regexp.MustCompile("(tags.)+(.)"),
+		},
+	}
+)
+
+// GetBreakingChanges compares each change in a diff Proto to the relavant change type detection Patterns.
+// Each change is then catgorized as breaking, nonbreaking, or unknown.
+func GetBreakingChanges(diff *rpc.Diff) *rpc.ClassifiedChanges {
+	changesProto := compareChangesToPatterns(diff)
+	return changesProto
+}
+
+func compareChangesToPatterns(diff *rpc.Diff) *rpc.ClassifiedChanges {
+	allChanges := &rpc.ClassifiedChanges{
+		BreakingChanges: &rpc.Diff{
+			Additions:     []string{},
+			Deletions:     []string{},
+			Modifications: make(map[string]*rpc.Diff_ValueChange),
+		},
+		NonBreakingChanges: &rpc.Diff{
+			Additions:     []string{},
+			Deletions:     []string{},
+			Modifications: make(map[string]*rpc.Diff_ValueChange),
+		},
+		UnknownChanges: &rpc.Diff{
+			Additions:     []string{},
+			Deletions:     []string{},
+			Modifications: make(map[string]*rpc.Diff_ValueChange),
+		},
+	}
+	allChanges.BreakingChanges = getBreakingChanges(diff)
+	allChanges.NonBreakingChanges = getNonBreakingChanges(diff)
+	allChanges.UnknownChanges = getUnknownChanges(diff)
+	return allChanges
```

Comment:
Now this function can be reduced to the following. 

```
return &rpc.ClassifiedChanges{
	BreakingChanges: getBreakingChanges(diff),
	NonBreakingChanges: getNonBreakingChanges(diff),
	UnknownChanges: getUnknownChanges(diff),
}
```

Since the function is only used in `GetBreakingChanges` you can actually remove the function entirely and return it all directly.

```
func GetBreakingChanges(diff *rpc.Diff) *rpc.ClassifiedChanges {
	return &rpc.ClassifiedChanges{
		BreakingChanges: getBreakingChanges(diff),
		NonBreakingChanges: getNonBreakingChanges(diff),
		UnknownChanges: getUnknownChanges(diff),
	}
}
```

Reply:
Done